run.go

```diff
@@ -139,38 +135,102 @@ var commandInspect = cli.Command{
 	Action: actionInspect,
 }
 
-func guessType(filename string) string {
-	switch {
-	case strings.Contains(filename, "://"):
-		return TypeURL
-	case strings.HasSuffix(filename, ".js"):
-		return TypeJS
-	default:
-		return ""
+func looksLikeURL(str []byte) bool {
+	// Try to parse string as URL
+	u, err := url.Parse(strings.TrimSpace(string(str)))
+	if err != nil || u.Scheme == "" {
+		return false
 	}
+	return true
 }
 
-func makeRunner(filename, t string) (lib.Runner, error) {
-	if t == TypeAuto {
-		t = guessType(filename)
+func getSrcData(arg, t string) (*lib.SourceData, error) {
+	srcdata := []byte("")
+	srctype := t
+	filename := arg
+	var err error
+	const cmdline = "[cmdline]"
+	// special case name “-“ will always cause src data to be read from file STDIN
+	if arg == "-" {
+		srcdata, err = ioutil.ReadAll(os.Stdin)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		// Deduce how to get src data
+		switch srctype {
+		case lib.TypeAuto:
+			if looksLikeURL([]byte(arg)) { // always try to parse as URL string first
+				srcdata = []byte(arg)
+				srctype = lib.TypeURL
+				filename = cmdline
+			} else {
+				// Otherwise, check if it is a file name and we can load the file
+				srcdata, err = ioutil.ReadFile(arg)
+				if err != nil { // if we fail to open file, we assume the arg is JS code
+					srcdata = []byte(arg)
+					srctype = lib.TypeJS
+					filename = cmdline
+				}
+			}
+		case lib.TypeURL:
+			// We try to use lib.TypeURL args as URLs directly first
+			if looksLikeURL([]byte(arg)) {
+				srcdata = []byte(arg)
+				filename = cmdline
+			} else { // if that didn’t work, we try to load a file with URLs
+				srcdata, err = ioutil.ReadFile(arg)
+				if err != nil {
+					return nil, err
+				}
+			}
+		case lib.TypeJS:
+			// lib.TypeJS args we try to use as file names first
+			srcdata, err = ioutil.ReadFile(arg)
+			if err != nil { // and if that didn’t work, we assume the arg itself is JS code
+				srcdata = []byte(arg)
+				filename = cmdline
+			}
```

Comment:
Maybe add a default case here just in case an invalid type is returned. Shouldn't happen, but it's better to give a clear error than to do the wrong thing if someone changes something internally and forgets to update this part.